n-api: Enable scope and ref APIs during exception

jasongin opened the pull request:
N-API is somewhat strict about blocking calls to many APIs while there is a pending exception. The `NAPI_PREAMBLE` macro at the beginning of many API implementations checks for a pending exception. However, a subset of the APIs (which don't call back into JavaScript) still need to work while in a pending-exception state. This changes the reference APIs (equivalent to `v8::Persistent`) and handle scope APIs so that they can be used for cleanup up while an exception is pending.

We may decide to similarly enable a few other APIs later, (which would be a non-breaking change) but we know at least these are needed now to unblock some specific scenarios.

There currently isn't good test coverage of the reference APIs; I'm working on that separately. This change only adds a handle scope test case.

Also fixing a couple places where I noticed the argument validation was incorrect.

Fixes: https://github.com/nodejs/abi-stable-node/issues/122

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

jasongin commented:
@mhdawson @gabrielschulhof please review.

gabrielschulhof commented on src/node_api.cc:
```diff
@@ -2011,14 +2012,16 @@ napi_status napi_create_reference(napi_env env,
 
   *result = reinterpret_cast<napi_ref>(reference);
-  return GET_RETURN_STATUS(env);
+  return napi_ok;
```
`NAPI_PREAMBLE(env)` clears the last error, whereas `CHECK_ENV(env)` does not. So, the public APIs which do not start with `NAPI_PREAMBLE(env)` leave the last error status as it was if they simply return `napi_ok`. Thus, I was thinking we should change `napi_clear_last_error(env)` to always return `napi_ok` and then change all these `return napi_ok;` statements to `return napi_clear_last_error(env);`

gabrielschulhof commented:
... and also LGTM because `return napi_ok;` is what we do in all our non-`NAPI_PREAMBLE(env)` APIs.

jasongin commented:
This will need to be merged with https://github.com/nodejs/node/pull/12539. It might be better to let that one go first, then I can rebase and update this PR to use the new patterns from that one.

mhdawson commented:
Ok, I'll wait and land the other one first.

jasongin commented:
Rebased, and added a commit to conform to the updated pattern in #12539 where we return `napi_clear_last_error()` from functions that don't use `NAPI_PREAMBLE`.

addaleax commented:
Fresh CI: https://ci.nodejs.org/job/node-test-commit/9425/

jasongin commented:
Oh, this also fixes https://github.com/nodejs/abi-stable-node/issues/228

addaleax commented:
Landed in b7a341d7e5868cd1279c33e9f62c388a1e886110 and added the extra Fixes: tag, thanks!
